tests/core/data/test_network.py

```diff
@@ -0,0 +1,334 @@
+    node = random_node((2,2))
+    network = Network(node[0:1], node[1:], copy=False)
+
+    copied_network = network.copy()
+    assert copied_network.in_edges[0] is not network.in_edges[0]
+    assert copied_network.out_edges[0] is not network.out_edges[0]
+
+    assert copied_network.nodes is not network.nodes
+    node = network.nodes.pop()
+    copied_node = copied_network.nodes.pop()
+    assert node is not copied_node
+    assert node.tensor is copied_node.tensor
+
+def test_fast_constructor():
+    node = random_node((2,2))
+    network = Network(node[0:1], node[1:], copy=False)
+
+    new_network = Network._fast_constructor(network.out_edges,
+                                             network.in_edges,
+                                             network.nodes,
+                                           )
+
+    assert network.in_edges is new_network.in_edges
+    assert network.out_edges is new_network.out_edges
+    assert network.nodes is new_network.nodes
+    assert network.shape == new_network.shape
+
+def test_shape():
+    node = random_node((2,2))
+
+    network = Network(node[0:], [])
+    assert network.shape == (4, 1)
+
+    network = Network([], node[0:])
+    assert network.shape == (1, 4)
+
+    network = Network([node[0]], [node[1]])
+    assert network.shape == (2, 2)
+
+# This are very simple tests that are not meant to be complete. For a complete
+# set of tests for the mathematical operations see test_mathematics. Both shape
+# and dims are tested as these are generated independently.
+def test_conj():
+    node = random_node((2,3))
+    network = Network([node[0]], [node[1]])
+
+    assert network.shape == network.conj().shape
+    assert network.dims == network.conj().dims
+
+    np.testing.assert_almost_equal(node.tensor.conj(),
+                                   network.conj().nodes.pop().tensor)
+
+def test_adjoint():
+    node = random_node((2,3))
+    network = Network([node[0]], [node[1]])
+
+    assert network.shape[0] == network.adjoint().shape[1]
+    assert network.shape[1] == network.adjoint().shape[0]
+    assert network.dims[0] == network.adjoint().dims[1]
+    assert network.dims[1] == network.adjoint().dims[0]
+
+    np.testing.assert_almost_equal(node.tensor.conj(),
+                                   network.adjoint().nodes.pop().tensor)
+
+def test_transpose():
+    node = random_node((2,3))
+    network = Network([node[0]], [node[1]])
+
+    assert network.shape[0] == network.transopose().shape[1]
+    assert network.shape[1] == network.transopose().shape[0]
+    assert network.dims[0] == network.transopose().dims[1]
+    assert network.dims[1] == network.transopose().dims[0]
+
+    np.testing.assert_almost_equal(node.tensor,
+                                   network.transopose().nodes.pop().tensor)
+
+class TestContract():
+    node1 = random_node((2,))
+    node2 = random_node((2, 2, 2))
+    node1[0] ^ node2[2]
+
+    network = Network([node2[0]], [node2[1]], [node1, node2])
```

Comment:
We should also just create this network within the individual tests if we can.

Reply:
I included the node and network creation inside the methods.